services/datamanager/builtin/file_deletion_test.go

```diff
@@ -204,43 +206,20 @@ func TestFilePolling(t *testing.T) {
 	flusher.closeCollectors()
 	// number of capture files is based on the number of unique
 	// collectors in the robot config used in this test
-	waitForCaptureFilesToExceedNFiles(tempDir, 4)
+	waitForCaptureFilesToEqualNFiles(tempDir, 4)
 
 	files := getAllFileInfos(tempDir)
 	test.That(t, len(files), test.ShouldEqual, 4)
 	// since we've written 4 files and hit the threshold, we expect
 	// the first to be deleted
 	expectedDeletedFile := files[0]
 
+	// run forward 20ms to delete any files
 	mockClock.Add(filesystemPollInterval)
+	waitForCaptureFilesToEqualNFiles(tempDir, 3)
 	newFiles := getAllFileInfos(tempDir)
 	test.That(t, len(newFiles), test.ShouldEqual, 3)
 	test.That(t, newFiles, test.ShouldNotContain, expectedDeletedFile)
-	// capture interval is 10ms and sync interval is 50ms.
-	// So we run forward 10ms to capture 4 files then close the collectors,
-	// run forward 20ms to delete any files, then run forward 20ms to sync the remaining 3 files.
-	mockClock.Add(syncInterval - (filesystemPollInterval + captureInterval))
-
-	wait := time.After(time.Second)
-	// we track requests to make sure only 3 upload requests are in the syncer
-	// instead of the 4 captured files
-	var requests []*v1.DataCaptureUploadRequest
-	select {
-	case req := <-mockClient.succesfulDCRequests:
-		requests = append(requests, req)
-	case <-wait:
-	}
-	flusher.closeSyncer()
-
-	close(mockClient.succesfulDCRequests)
-	for dc := range mockClient.succesfulDCRequests {
-		requests = append(requests, dc)
-	}
-	// we expect 3 upload requests as 1 capture file should have been deleted
-	// and not uploaded
-	test.That(t, len(requests), test.ShouldEqual, 3)
```

Comment:
It's impossible to make this code correct with the internals/API that the data sync Manager provides right now. This assertion was failing in practice for 2 distinct reasons:
- The "[bump the clock](https://github.com/viamrobotics/rdk/pull/3950/files#diff-81546ab6903654498a5f19b6d20b0a70bb2170243db5a9aa5bd5beef14aa33e9L222)" to kick off data sync also kicked off a second deletion job. If the second deletion job won, we'd see 2 files getting synced instead of 3.
- Datasync spins off 1 goroutine for each of the 3 files to sync. We [close](https://github.com/viamrobotics/rdk/pull/3950/files#diff-81546ab6903654498a5f19b6d20b0a70bb2170243db5a9aa5bd5beef14aa33e9L233) the syncer after seeing [one goroutine complete](https://github.com/viamrobotics/rdk/pull/3950/files#diff-81546ab6903654498a5f19b6d20b0a70bb2170243db5a9aa5bd5beef14aa33e9L229). There's no guarantee the other goroutines were [started/added to the waitgroup](https://github.com/viamrobotics/rdk/blob/main/services/datamanager/datasync/sync.go#L132).

I fixed the first, but had to make an invasive change to fix the second. I think given we're asserting three files remain in the data capture directory after deletion, we don't need to write more assertions in this specific test regarding datasync's ability to also finding 3 files.

Reply:
Really great digging and solutions here. All of this makes sense, and should have noticed these concurrency issues when reviewing initially.

> I think given we're asserting three files remain in the data capture directory after deletion, we don't need to write more assertions in this specific test regarding datasync's ability to also finding 3 files.

Agreed - it was a nice to have, but already covered by datasync tests.